sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/GotoStatementBase.cs

```diff
@@ -0,0 +1,54 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2020 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class GotoStatementBase<TLanguageKindEnum> : SonarDiagnosticAnalyzer
+        where TLanguageKindEnum : struct
+    {
+        protected const string DiagnosticId = "S907";
+        internal const string MessageFormat = "Remove this use of '{0}'.";
+        private readonly DiagnosticDescriptor rule;
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+        protected abstract TLanguageKindEnum[] GotoSyntaxKinds { get; }
+        protected abstract string GoToLabel { get; }
+
+        protected GotoStatementBase(System.Resources.ResourceManager rspecResources)
+        {
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources);
```

Comment:
`GoToLabel` value does not depend on the source code itself. In order to simplify dependencies, it would be nice to have `gotoLabel` as argument here and build the descriptor directly.
```suggestion
            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, string.Format(MessageFormat, gotoLabel), rspecResources);
```

Reply:
What I exactly meant was:
- Remove `GoToLabel` abstract method
- Add `gotoLabel` argument to constructor
- Update inherited classes accordingly to simplify class structure and dependencies.

Invoking inherited abstract method from constructor is risky pattern.